docs: finishing chain-indexer-overview-for-operators.md (#12600)
filecoin-project/lotus: documentation/en/chain-indexer-overview-for-operators.md

  - [`ChainValidateIndex` RPC API](#chainvalidateindex-rpc-api)

## Introduction

This document is for externally-available and/or high-performance RPC providers.  It walks through the configuration changes, migration flow and operations/maintenance work needed to enable, backfill and maintain the [`ChainIndexer`](#chainindexer-indexing-system).  The justification for and benefits of the `ChainIndexer` are documented [here](https://github.com/filecoin-project/lotus/issues/12453). 

The ChainIndexer is now also required if you enable Ethereum (`eth_*`) APIs using the `EnableEthRPC` Lotus configuration option.

**Note: If you are a Storage Provider or node operator who does not serve public RPC requests or does not need Ethereum APIs enabled (i.e, if `Fevm.EnableEthRPC = false`), you can skip this document as the `ChainIndexer` is already disabled by default**. 

## ChainIndexer Config
### Enablement

The following must be enabled on an Lotus node before starting as they are disabled by default:

```toml
[Fevm]
# Enable the ETH RPC APIs.
# This is not required for ChainIndexer support, but ChainIndexer is required if you enable this.
  EnableEthRPC = true

[Events]

> # Enable the Actor Events APIs.

Reviewer: ```suggestion
# Enable the Actor Events APIs.
# This is not required for ChainIndexer support, but ChainIndexer is required if you enable this.
```

Suggesting this since it's the same argument as for `EnableEthRPC`: https://github.com/filecoin-project/lotus/pull/12421/files#r1803857863

Even if we don't make it mandatory, this setting isn't a "must" for ChainIndexer, it just enables an optional feature that uses ChainIndexer.

**Warning:** Setting this value below the chain state retention period may degrade RPC performance and reliability because the ChainIndexer will lack data for epochs still present in the chain state.

**Note:** `Chainstore.Splitstore` is configured in terms of bytes (not epochs) and `ChainIndexer.GCRetentionEpochs` is in terms of epochs (not bytes).  For the purposes of this discussion, we're assuming operators have determined `Chainstore.Splitstore.HotStoreMaxSpaceTarget` and `Chainstore.Splitstore.HotStoreMaxSpaceThreshold` values that approximate a certain number days of storage in the Splitstore Hotstore.  The guidance here is to make sure this approximation exceeds `ChainIndexer.GCRetentionEpochs`.

### Removed Options

**Note: The following config options no longer exist in Lotus and have been removed in favor of the ChainIndexer config options explained above. They can be removed when upgrading to Lotus v1.31.0.**

```toml
[Fevm]
EthTxHashMappingLifetimeDays

> DisableHistoricFilterAPI

Reviewer: ```suggestion
```

sorry, this one is in Fevm.Events and is covered in the list below

**Warning:** Setting this value below the chain state retention period may degrade RPC performance and reliability because the ChainIndexer will lack data for epochs still present in the chain state.

**Note:** `Chainstore.Splitstore` is configured in terms of bytes (not epochs) and `ChainIndexer.GCRetentionEpochs` is in terms of epochs (not bytes).  For the purposes of this discussion, we're assuming operators have determined `Chainstore.Splitstore.HotStoreMaxSpaceTarget` and `Chainstore.Splitstore.HotStoreMaxSpaceThreshold` values that approximate a certain number days of storage in the Splitstore Hotstore.  The guidance here is to make sure this approximation exceeds `ChainIndexer.GCRetentionEpochs`.

### Removed Options

**Note: The following config options no longer exist in Lotus and have been removed in favor of the ChainIndexer config options explained above. They can be removed when upgrading to Lotus v1.31.0.**

```toml
[Fevm]
EthTxHashMappingLifetimeDays

> [Events]
> DisableHistoricFilterAPI

Reviewer: ```suggestion
[Events]
DisableRealTimeFilterAPI
DisableHistoricFilterAPI
```

I'm proposing even more simplification: https://github.com/filecoin-project/lotus/pull/12610

    - [Part 2: Create a copyable `chainindex.db`](#part-2-create-a-copyable-chainindexdb)
    - [Part 3: Update other nodes](#part-3-update-other-nodes)
    - [Part 4: Cleanup](#part-4-cleanup)
  - [Upgrade when importing chain state from a snapshot](#upgrade-when-importing-chain-state-from-a-snapshot)
- [Backfill](#backfill)
  - [Backfill Timing](#backfill-timing)
  - [Backfill Disk Space Requirements](#backfill-disk-space-requirements)
  - [`lotus-shed chainindex validate-backfill` CLI tool](#lotus-shed-chainindex-validate-backfill-cli-tool)
    - [Usage](#usage)
- [Regular Checks](#regular-checks)
- [Downgrade Steps](#downgrade-steps)
- [Terminology](#terminology)
  - [Previous Indexing System](#previous-indexing-system)
  - [ChainIndexer Indexing System](#chainindexer-indexing-system)
- [Appendix](#appendix)
  - [Wny isn't there an automated migration from the previous indexing system to the ChainIndexer indexing system?](#wny-isnt-there-an-automated-migration-from-the-previous-indexing-system-to-the-chainindexer-indexing-system)
  - [`ChainValidateIndex` RPC API](#chainvalidateindex-rpc-api)

## Introduction

> This document is for externally-available and/or high-performance RPC providers.  It walks through the configuration changes, migration flow and operations/maintenance work needed to enable, backfill and maintain the [`ChainIndexer`](#chainindexer-indexing-system).  The justification for and benefits of the `ChainIndexer` are documented [here](https://github.com/filecoin-project/lotus/issues/12453). 
> 
> The ChainIndexer is now also required if you enable Ethereum (`eth_*`) APIs using the `EnableEthRPC` Lotus configuration option.
> 
> **Note: If you are a Storage Provider or node operator who does not serve public RPC requests or does not need Ethereum APIs enabled (i.e, if `Fevm.EnableEthRPC = false`), you can skip this document as the `ChainIndexer` is already disabled by default**. 

Reviewer: ```suggestion
This document is for externally-available, high-performance RPC providers and for node operators who use or expose the Ethereum and/or events APIs.  It walks through the configuration changes, migration flow and operations/maintenance work needed to enable, backfill and maintain the [`ChainIndexer`](#chainindexer-indexing-system).  The justification for and benefits of the `ChainIndexer` are documented [here](https://github.com/filecoin-project/lotus/issues/12453). 

The ChainIndexer is now also required if you enable Ethereum (`eth_*`) APIs using the `EnableEthRPC` Lotus configuration option.

**Note: If you are a Storage Provider or node operator who does not serve public RPC requests or does not need Ethereum or Event APIs (i.e, if `Fevm.EnableEthRPC = false` and `Events.EnableActorEventsAPI = false`, their default values), you can skip this document as the `ChainIndexer` is already disabled by default**. 
```

## ChainIndexer Config
### Enablement

The following must be enabled on an Lotus node before starting as they are disabled by default:

```toml
[Fevm]
# Enable the ETH RPC APIs.
# This is not required for ChainIndexer support, but ChainIndexer is required if you enable this.
  EnableEthRPC = true

[Events]
# Enable the Actor Events APIs.
# This is not required for ChainIndexer support, but ChainIndexer is required if you enable this.
  EnableActorEventsAPI = true

[ChainIndexer]
# Enable the ChainIndexer, which is required for the ETH RPC APIs and Actor Events APIs.
# If they are enabled, but the ChainIndexer is not, Lotus will exit during startup.

> # (ChainIndexer needs to be explicitly enabled to signal to the node operator the extra supporting functionalithy that will now be running.)

Reviewer: ```suggestion
# (ChainIndexer needs to be explicitly enabled to signal to the node operator the extra
# supporting functionalithy that will now be running.)
```

Reviewer: functionalithy, it's like functionality, just filthy

    - [Usage](#usage)
- [Regular Checks](#regular-checks)
- [Downgrade Steps](#downgrade-steps)
- [Terminology](#terminology)
  - [Previous Indexing System](#previous-indexing-system)
  - [ChainIndexer Indexing System](#chainindexer-indexing-system)
- [Appendix](#appendix)
  - [Wny isn't there an automated migration from the previous indexing system to the ChainIndexer indexing system?](#wny-isnt-there-an-automated-migration-from-the-previous-indexing-system-to-the-chainindexer-indexing-system)
  - [`ChainValidateIndex` RPC API](#chainvalidateindex-rpc-api)

## Introduction

This document is for externally-available, high-performance RPC providers and for node operators who use or expose the Ethereum and/or events APIs.  It walks through the configuration changes, migration flow and operations/maintenance work needed to enable, backfill and maintain the [`ChainIndexer`](#chainindexer-indexing-system).  The justification for and benefits of the `ChainIndexer` are documented [here](https://github.com/filecoin-project/lotus/issues/12453). 

The ChainIndexer is now also required if you enable Ethereum (`eth_*`) APIs using the `EnableEthRPC` Lotus configuration option.

> 

Reviewer 2: In addition -> `ChainIndexer` is ALSO required if user enables the ActorEventAPIs using the `EnableActorEventsAPI` using the Lotus configuration option

Author: Fixed in https://github.com/filecoin-project/lotus/pull/12600/commits/8463a03345d8c28410825d0cc115224976522d31.  Resolving.

The ChainIndexer [periodically runs](https://github.com/filecoin-project/lotus/blob/master/chain/index/gc.go#L15) GC if `GCRetentionEpochs` is > 0 and removes indexed data for epochs older than `(current_head_height - GCRetentionEpochs)`.

```toml
[ChainIndexer]
  GCRetentionEpochs = X  # Replace X with your desired value
```

- Setting `GCRetentionEpochs` to 0 (**default**) disables GC.
- Any positive value enables GC and determines the number of epochs of historical data to retain.

#### Recommendations

1. **Archival Nodes**: **Keep GC disabled** (`GCRetentionEpochs` = 0) to retain all indexed data.

2. **Non-Archival Nodes**:  Set `GCRetentionEpochs` to match the amount of chain state your node retains 

**Example:** if your node is configured to retain Filecoin chain state with a Splitstore Hotstore that approximates 2 days of epochs, set `GCRetentionEpochs` to at least `retentionDays * epochsPerDay = 2 * 2880 = 5760`).

**Warning:** Setting this value below the chain state retention period may degrade RPC performance and reliability because the ChainIndexer will lack data for epochs still present in the chain state.

> **Note:** `Chainstore.Splitstore` is configured in terms of bytes (not epochs) and `ChainIndexer.GCRetentionEpochs` is in terms of epochs (not bytes).  For the purposes of this discussion, we're assuming operators have determined `Chainstore.Splitstore.HotStoreMaxSpaceTarget` and `Chainstore.Splitstore.HotStoreMaxSpaceThreshold` values that approximate a certain number days of storage in the Splitstore Hotstore.  The guidance here is to make sure this approximation exceeds `ChainIndexer.GCRetentionEpochs`.

Reviewer 2: @rvagg @BigLep 

Have we checked the logic related to `HotStoreMaxSpaceTarget` and `HotStoreMaxSpaceThreshold` to be absolutely certain of what we're talking about here ? 

Unless we're **_absolutely_** certain that this is the only splitstore config we're trying to approximate in `GCRetentionEpochs` -> I honestly think this section will confuse RPC providers using (snapshots + splitstore) more than helping them.

Will defer the final decision to you'll as I do not know enough about the splitstore but I do know that RPC providers also do not know much about the splitstore. Please do give it a thought.

Reviewer: All I know is that `GCRetentionEpochs` doesn't have a direct relationship to the way splitstore works so we shouldn't mislead in specifying a timeframe that matches splitstore.

Perhaps @ZenGround0 could help us here—would you mind sanity checking this line, do you think it's a reasonable statement given your knowledge of splitstore?

Reviewer 2: @rvagg 

So why not just say "make sure the GCRetentionEpochs epoch should be in line with how much chain state is retained on your node/how long back in the past RPC queries are served for".

RPC providers have good operational expertise and know things such as "My node only serves requests for the past 16 hours of epochs" or "My node only retains 3 days of chain state" and so can align the `GCRetentionEpochs` param with that.

I really don't think we should be taking them down a journey of `HotStoreMaxSpaceTarget ` etc etc here -> it will confuse them as I'm fairly confident they wont know what to do with this information.

Author: I don't have strong feelings here so will default to you guys.  That said, with a free re-read, I am still good with what's written.  This is because it helps draw attention that:
1. there are multiple configuration knobs that a splitstore/hostore-using operator needs to pay attention to AND 
2. that splitstore/hotstore configuration is in terms of bytes but ChainIndexer configuration is in terms of epochs

I believe those are the main points to drive home.

I agree that splitstore config and operations feel complex as I read https://github.com/filecoin-project/lotus/blob/master/documentation/en/default-lotus-config.toml#L151 , but after I got my head into the mental model of how it works based on the comments there, I think calling out `Chainstore.Splitstore.HotStoreMaxSpaceTarget` and `Chainstore.Splitstore.HotStoreMaxSpaceThreshold` seems reasonable.  (If the code doesn't match what's described in the comments, then I don't know what to say...)

Anyways, I will leave this as is unless you guys insist otherwise or we get operator feedback on this.  (Feel free to make changes too.)

### Removed Options

**Note: The following config options no longer exist in Lotus and have been removed in favor of the ChainIndexer config options explained above. They can be removed when upgrading to Lotus v1.31.0.**

```toml
[Fevm]
EthTxHashMappingLifetimeDays

[Events]
DisableRealTimeFilterAPI
DisableHistoricFilterAPI
DatabasePath

[Index]
EnableMsgIndex
```

The previously deprecated `Fevm.Events` options are now also all removed:
* `Fevm.Events.DatabasePath`
* `Fevm.Events.DisableRealTimeFilterAPI`
* `Fevm.Events.DisableHistoricFilterAPI`

> * `Fevm.Events.FilterTTL`

Reviewer 2: The `FilterTTL`, `MaxFilters`, `MaxFilterResults` AND `MaxFilterHeightRange` options have NOT been removed and are still applicable as the `ChainIndexer` does not make any changes to the ETH Event Filter APIs.

Let's fix this.

Reviewer: They've been removed from `Fevm.Events`, that's the deprecated (`moved`) block that got deleted.

Reviewer 2: Ah okay, they do still exist in `EventsConfig`. 

So might be worth calling out that the **_only_** way to configure these now might be `EventsConfig. FilterTTL`, `EventsConfig.MaxFilters` etc .

Author: I don't know the differences enough to write this.  Can you make a proposal or change @aarshkshah1992 ?  (or have quick sync with me so I can come up to speed)

Author: I think I fixed this in https://github.com/filecoin-project/lotus/pull/12600/commits/8463a03345d8c28410825d0cc115224976522d31 .  That said, I'm confused why we support it in one area and not in the other.

Or is this a case that these are good config options but they never should have been exposed through `Fevm.Events` and we're now cleaning that up?  

I don't know enough here, so please feel free to take over the wording here.

Reviewer: @BigLep they were all originally only on `Fevm.Events` but I moved them to a new top-level `Events` when we introduced builtin actor events. At that time, instead of hard-removing them, I introduced a soft deprecation method, where I deprecated them in documentation so they don't show up and they have a "moved" tag associated with them so when someone starts their node using one of these settings, it sets the right one automatically for them and complains in their logs that it's deprecated and they should fix it.

So with this release, we're deciding to hard-deprecate and just remove them, no more complaints, no more automatic setting of the right value. I'll add a suggestion to clear it up a little more.

## Backfill
There is no automated migration from [pre-ChainIndexer indices](#previous-indexing-system) to the [ChainIndex](#chainindexer-indexing-system).  Instead one needs to index historical chain state (i.e., backfill), if RPC access to that historical state is required. (If curious, [read why](#wny-isnt-there-an-automated-migration-from-the-previous-indexing-system-to-the-chainindexer-indexing-system).)

### Backfill Timing

Backfilling the new `ChainIndexer` was [benchmarked to take approximately ~12 hours per month of epochs on a sample archival node doing no other work](https://github.com/filecoin-project/lotus/issues/12453#issuecomment-2405306468). Your results will vary depending on hardware, network, and competing processes.  This means if one is upgrading a FEVM archival node, they should plan on the node being out of production service for ~10 days.  Additional nodes to update don't need to go throuh the same time-intensive process though.  They can get a `${LOTUS_PATH}/chainindex/chainindex.db` copied from a trusted node per the [upgrade steps](#upgrade).  

### Backfill Disk Space Requirements

> As of 202410, ChainIndexer will accumulate approximately ~340 MiB per day of data, or 10 GiB per month (see [here](https://github.com/filecoin-project/lotus/issues/12453)).

Reviewer 2: This is correct. See https://filecoinproject.slack.com/archives/CP50PPW2X/p1729413621133599.

* It persisted state to the [removed option](#removed-options) for `Events.DatabasePath`, which defaulted to `${LOTUS_PATH}/sqlite`.
* It had CLI backfill tooling: `lotus-shed index backfill-*`

### ChainIndexer Indexing System
* This corresponds to the indexing system used in Lotus versions v1.31.0 onwards.  
* It replaced the [previous indexing system](#previous-indexing-system).
* It is composed of a single indexer, [`ChainIndexer`](https://github.com/filecoin-project/lotus/blob/master/chain/index/indexer.go), using a [single database for transactions, messages, and events](https://github.com/filecoin-project/lotus/blob/master/chain/index/ddls.go).
* It persists state to `${LOTUS_PATH}/chainindex`.
* It has this CLI backfill tooling: [`lotus-shed chainindex validate-backfill`](#lotus-shed-chainindex-validate-backfill-cli-tool)
* **Storage requirements:** See the [backfill disk space requirements](#backfill-disk-space-requirements).
* **Backfil times:** See the [backfill timing](#backfill-timing).

## Appendix

> ### Wny isn't there an automated migration from the [previous indexing system](#previous-indexing-system) to the [ChainIndexer indexing system](#chainindexer-indexing-system)?

Reviewer 2: Why*

Author: Fixed in https://github.com/filecoin-project/lotus/pull/12600/commits/8463a03345d8c28410825d0cc115224976522d31 .  Resolving.

**Warning:** Setting this value below the chain state retention period may degrade RPC performance and reliability because the ChainIndexer will lack data for epochs still present in the chain state.

**Note:** `Chainstore.Splitstore` is configured in terms of bytes (not epochs) and `ChainIndexer.GCRetentionEpochs` is in terms of epochs (not bytes).  For the purposes of this discussion, we're assuming operators have determined `Chainstore.Splitstore.HotStoreMaxSpaceTarget` and `Chainstore.Splitstore.HotStoreMaxSpaceThreshold` values that approximate a certain number days of storage in the Splitstore Hotstore.  The guidance here is to make sure this approximation exceeds `ChainIndexer.GCRetentionEpochs`.

### Removed Options

**Note: The following config options no longer exist in Lotus and have been removed in favor of the ChainIndexer config options explained above. They can be removed when upgrading to Lotus v1.31.0.**

```toml
[Fevm]
EthTxHashMappingLifetimeDays

[Events]
DisableRealTimeFilterAPI
DisableHistoricFilterAPI
DatabasePath

[Index]
EnableMsgIndex
```

> The previously deprecated `Fevm.Events` options are now also all removed:
> * `Fevm.Events.DatabasePath`
> * `Fevm.Events.DisableRealTimeFilterAPI`
> * `Fevm.Events.DisableHistoricFilterAPI`
> * `Fevm.Events.FilterTTL`
> * `Fevm.Events.MaxFilters`
> * `Fevm.Events.MaxFilterResults`
> * `Fevm.Events.MaxFilterHeightRange`
> They do still exist in `EventsConfig` and can be configured there.

Reviewer: ```suggestion
The `Fevm.Events` options were marked as deprecated in Lotus 1.26, having been moved to the new top-level `Events` section, and have now been removed with Lotus 1.31.

* `Fevm.Events.DatabasePath` (no replacement available)
* `Fevm.Events.DisableRealTimeFilterAPI` (no replacement available)
* `Fevm.Events.DisableHistoricFilterAPI` (no replacement available)
* `Fevm.Events.FilterTTL` (use `Events.FilterTTL` intead)
* `Fevm.Events.MaxFilters` (use `Events.MaxFilters` instead)
* `Fevm.Events.MaxFilterResults` (use `Events.MaxFilterResults` instead)
* `Fevm.Events.MaxFilterHeightRange` (use `Events.MaxFilterHeightRange` instead)
```
